ufl/action.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+"""This module defines the Matrix class."""
+
+# Copyright (C) 2021 India Marsden
+#
+# This file is part of UFL (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+
+from ufl.form import BaseForm, FormSum, Form
+from ufl.coefficient import Coefficient
+
+# --- The Action class represents the adjoint of a numerical object that needs to be computed at compile time ---
+
+
+class Action(BaseForm):
```

Comment:
Also, at the moment `Action(F, x)` assumes that `x` has arguments, is there a reason for that ?

They are cases where this is not true: for example, if `action(F, x)` is called with `x` a `Coefficient` and `F` a `BaseForm` that is not a `Form` (e.g. if Action is called twice: `F = Action(...)`) then we end up with an error since `Action` assumes that `_right` has arguments (this situation actually occurs for external operators).

Reply:
I think Action checks the type of right before assuming it has arguments. If right is a Form, then it is defined to have arguments. if right is a coefficient, then we only look at the arguments of left, which are required. I just wrote up a test for your example of calling action twice, and it does work as it filters into the case where right is a coefficient both times. (left is F = Action(..), right is x = Coefficient). I'll have a think about the zero case.